File: mcp_use/observability/langfuse.py
import logging
import os
from typing import Any

from .types import LangfuseObservabilityConfig

logger = logging.getLogger(__name__)

# Global instances
_langfuse_client = None
_current_config: LangfuseObservabilityConfig | None = None

# Expose patchable references for tests
try:  # pragma: no cover
    from langfuse import Langfuse as Langfuse  # type: ignore
except Exception:  # pragma: no cover
    Langfuse = None  # type: ignore

try:  # pragma: no cover
    from langfuse.langchain import CallbackHandler as CallbackHandler  # type: ignore
except Exception:  # pragma: no cover
    CallbackHandler = None  # type: ignore


def _is_langfuse_available() -> bool:
    """Check if Langfuse is available for import."""
    return Langfuse is not None and CallbackHandler is not None


def _has_required_env_vars() -> bool:
    """Check if required environment variables are set."""
    return bool(os.getenv("LANGFUSE_PUBLIC_KEY") and os.getenv("LANGFUSE_SECRET_KEY"))


def _is_disabled_by_env() -> bool:
    """Check if Langfuse is disabled via environment variable."""
    return os.getenv("MCP_USE_LANGFUSE", "").lower() == "false"


def configure_langfuse(config: LangfuseObservabilityConfig) -> None:
    """Configure Langfuse with the provided configuration.

    Args:
        config: Langfuse observability configuration
    """
    global _langfuse_client, _current_config

    _current_config = config

    # Reset client
    _langfuse_client = None

    if _is_disabled_by_env():
        logger.debug("Langfuse tracing disabled via MCP_USE_LANGFUSE environment variable")
        return

    if not config.is_enabled():
        logger.debug("Langfuse tracing disabled via configuration")
        return

    if not _has_required_env_vars():
        logger.debug(
            "Langfuse API keys not found - tracing disabled. Set LANGFUSE_PUBLIC_KEY and LANGFUSE_SECRET_KEY to enable"
        )
        return

    if not _is_langfuse_available():
        logger.debug("Langfuse package not installed - tracing disabled. Install with: pip install langfuse")
        return

    try:
        if Langfuse is None:
            raise RuntimeError("Langfuse class not available")

        _langfuse_client = Langfuse(
            public_key=os.getenv("LANGFUSE_PUBLIC_KEY"),
            secret_key=os.getenv("LANGFUSE_SECRET_KEY"),
            host=os.getenv("LANGFUSE_HOST", "https://cloud.langfuse.com"),
        )

        logger.info(f"✅ Langfuse client initialized successfully (trace_level: {config.trace_level})")
    except Exception as e:
        logger.debug(f"Failed to initialize Langfuse client: {e}")
        _langfuse_client = None


def get_langfuse_callbacks() -> list[Any]:
    """Get Langfuse callback handlers for LangChain integration.

    Returns:
        List of callback handlers, empty if Langfuse is not configured
    """
    if not _langfuse_client or not _current_config or not _current_config.is_enabled():
        logger.debug("🔍 No Langfuse callbacks: client not initialized or config disabled")
        return []

    if not _is_langfuse_available():
        logger.debug("🔍 No Langfuse callbacks: langfuse package not available")
        return []

    try:
        if CallbackHandler is None:
            raise RuntimeError("CallbackHandler not available")

        # Create callback handler with current configuration
        callback_kwargs = {}

        if _current_config.session_id:
            callback_kwargs["session_id"] = _current_config.session_id

        if _current_config.user_id:
            callback_kwargs["user_id"] = _current_config.user_id

        # Build metadata from configuration
        metadata = _current_config.metadata.copy()
        metadata.update(
            {
                "trace_level": _current_config.trace_level,
                "capture_tool_inputs": _current_config.capture_tool_inputs,
                "capture_tool_outputs": _current_config.capture_tool_outputs,
                "capture_context": _current_config.capture_context,
                "filter_sensitive_data": _current_config.filter_sensitive_data,
            }
        )

        if metadata:
            callback_kwargs["metadata"] = metadata

        handler = CallbackHandler(**callback_kwargs)
        logger.info(f"📊 Created Langfuse callback handler (session: {_current_config.session_id or 'default'})")
        return [handler]

    except Exception as e:
        logger.debug(f"Failed to create Langfuse callback handler: {e}")
        return []


def get_current_config() -> LangfuseObservabilityConfig | None:
    """Get the current Langfuse configuration."""
    return _current_config


def is_langfuse_enabled() -> bool:
    """Check if Langfuse is currently enabled and configured."""
    return _langfuse_client is not None and _current_config is not None and _current_config.is_enabled()


# Legacy compatibility - initialize with basic configuration if environment variables are present
def _initialize_legacy_compatibility():
    """Initialize Langfuse for backward compatibility if env vars are present."""
    global _langfuse_client, _current_config

    if _is_disabled_by_env():
        logger.debug("Langfuse tracing disabled via MCP_USE_LANGFUSE environment variable")
        return

    if not _has_required_env_vars():
        logger.debug(
            "Langfuse API keys not found - tracing disabled. Set LANGFUSE_PUBLIC_KEY and LANGFUSE_SECRET_KEY to enable"
        )
        return

    if not _is_langfuse_available():
        logger.debug("Langfuse package not installed - tracing disabled. Install with: pip install langfuse")
        return

    # Only initialize if no explicit configuration was provided
    if _current_config is None:
        default_config = LangfuseObservabilityConfig()
        configure_langfuse(default_config)
        logger.debug("Langfuse initialized with default configuration for backward compatibility")


# Legacy exports for backward compatibility
class _LegacyAccess:
    """Helper class for backward compatibility access patterns."""

    @property
    def langfuse(self):
        """Legacy property for accessing Langfuse client."""
        return _langfuse_client

    @property
    def langfuse_handler(self):
        """Legacy property for accessing Langfuse handler."""
        callbacks = get_langfuse_callbacks()
        return callbacks[0] if callbacks else None


# Create legacy access instance
_legacy = _LegacyAccess()

# Export for backward compatibility
langfuse = _legacy.langfuse
langfuse_handler = _legacy.langfuse_handler


# Initialize for backward compatibility
_initialize_legacy_compatibility()

File: libraries/typescript/packages/mcp-use/src/observability/langfuse.ts
/**
 * Langfuse observability integration for MCP-use.
 *
 * This module provides automatic instrumentation and callback handler
 * for Langfuse observability platform.
 *
 * Note: This module expects environment variables to be loaded before import.
 * Users should load their environment variables using their preferred method
 * (e.g., dotenv, direct process.env assignment, or system environment).
 */
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="./types.d.ts" />

import type { BaseCallbackHandler } from "@langchain/core/callbacks/base";
import type { Langfuse } from "langfuse";
import { logger } from "../logging.js";

type LangfuseMetadata = Record<string, unknown>;

interface LangfuseCallbackConfig extends LangfuseMetadata {
  verbose?: boolean;
}

/**
 * Retrieve the value of an environment variable when `process.env` is available.
 *
 * @param key - The environment variable name to look up
 * @returns The variable's value if present, `undefined` otherwise
 */
function getEnvVar(key: string): string | undefined {
  if (typeof process !== "undefined" && process.env) {
    return process.env[key];
  }
  return undefined;
}

function recordFromUnknown(value: unknown): LangfuseMetadata {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as LangfuseMetadata)
    : {};
}

function stringMetadataValue(
  metadata: LangfuseMetadata,
  key: string
): string | undefined {
  const value = metadata[key];
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

// Check if Langfuse is disabled via environment variable
const langfuseDisabled =
  getEnvVar("MCP_USE_LANGFUSE")?.toLowerCase() === "false";

// Initialize variables - using const with object to avoid linter issues with mutable exports
const langfuseState = {
  handler: null as BaseCallbackHandler | null,
  client: null as Langfuse | null,
  initPromise: null as Promise<void> | null,
};

/**
 * Initializes Langfuse observability for the application and installs a callback handler that augments traces with optional agent metadata and tags.
 *
 * This will attempt to dynamically load the Langfuse LangChain integration and, if available, create and store a wrapped callback handler (and optionally a Langfuse client) on the module state so tracing can be used elsewhere in the application.
 *
 * @param agentId - Optional identifier for the agent to include in traces
 * @param metadata - Optional static metadata to attach to traces; merged with dynamic metadata if a provider is supplied
 * @param metadataProvider - Optional function that returns dynamic metadata to attach to traces at runtime
 * @param tagsProvider - Optional function that returns an array of tags to attach to traces at runtime
 */
async function initializeLangfuse(
  agentId?: string,
  metadata?: LangfuseMetadata,
  metadataProvider?: () => LangfuseMetadata,
  tagsProvider?: () => string[]
): Promise<void> {
  try {
    // Dynamically import to avoid errors if package not installed
    const langfuseModule = await import("langfuse-langchain").catch(() => null);
    if (!langfuseModule) {
      logger.debug(
        "Langfuse package not installed - tracing disabled. Install with: npm install @langfuse/langchain"
      );
      return;
    }

    const { CallbackHandler } = langfuseModule;
    // Create a custom CallbackHandler wrapper to add logging and custom metadata
    class LoggingCallbackHandler extends CallbackHandler {
      private agentId?: string;
      private metadata?: LangfuseMetadata;
      private metadataProvider?: () => LangfuseMetadata;
      private tagsProvider?: () => string[];
      private isVerbose: boolean;

      constructor(
        config?: LangfuseCallbackConfig,
        agentId?: string,
        metadata?: LangfuseMetadata,
        metadataProvider?: () => LangfuseMetadata,
        tagsProvider?: () => string[]
      ) {
        super(config);
        this.agentId = agentId;
        this.metadata = metadata;
        this.metadataProvider = metadataProvider;
        this.tagsProvider = tagsProvider;
        this.isVerbose = config?.verbose ?? false;
      }

      // Override to add custom metadata to traces
      async handleChainStart(
        chain: unknown,
        inputs: unknown,
        runId?: string,
        parentRunId?: string,
        tags?: string[],
        metadata?: unknown,
        name?: string,
        kwargs?: unknown
      ): Promise<void> {
        logger.debug("Langfuse: Chain start intercepted");

        // Add custom tags and metadata
        const customTags = this.getCustomTags();
        const metadataToAdd = this.getMetadata();

        // Merge with existing tags and metadata
        const enhancedTags = [...(tags || []), ...customTags];
        const enhancedMetadata = {
          ...recordFromUnknown(metadata),
          ...metadataToAdd,
        };

        if (this.isVerbose) {
          logger.debug(
            `Langfuse: Chain start with custom tags: ${JSON.stringify(enhancedTags)}`
          );
          logger.debug(
            `Langfuse: Chain start with metadata: ${JSON.stringify(enhancedMetadata)}`
          );
        }

        return super.handleChainStart(
          chain,
          inputs,
          runId,
          parentRunId,
          enhancedTags,
          enhancedMetadata,
          name,
          kwargs
        );
      }

      // Get custom tags based on environment and agent configuration
      private getCustomTags(): string[] {
        const tags: string[] = [];

        // Add environment tag
        const env = this.getEnvironmentTag();
        if (env) {
          tags.push(`env:${env}`);
        }

        // Add agent ID tag if available
        if (this.agentId) {
          tags.push(`agent_id:${this.agentId}`);
        }

        // Add tags from provider if available
        if (this.tagsProvider) {
          const providerTags = this.tagsProvider();
          if (providerTags && providerTags.length > 0) {
            tags.push(...providerTags);
          }
        }

        return tags;
      }

      // Get metadata
      private getMetadata(): LangfuseMetadata {
        const metadata: LangfuseMetadata = {};

        // Add environment metadata
        const env = this.getEnvironmentTag();
        if (env) {
          metadata.env = env;
        }

        // Add agent ID metadata if available
        if (this.agentId) {
          metadata.agent_id = this.agentId;
        }

        // Add static metadata if provided
        if (this.metadata) {
          Object.assign(metadata, this.metadata);
        }

        // Add dynamic metadata from provider if available
        if (this.metadataProvider) {
          const dynamicMetadata = this.metadataProvider();
          if (dynamicMetadata) {
            Object.assign(metadata, dynamicMetadata);
          }
        }

        return metadata;
      }

      // Determine environment tag based on MCP_USE_AGENT_ENV
      private getEnvironmentTag(): string | null {
        const agentEnv = getEnvVar("MCP_USE_AGENT_ENV");
        if (!agentEnv) {
          // Default to 'unknown' if environment is not explicitly set
          return "unknown";
        }

        const envLower = agentEnv.toLowerCase();
        if (envLower === "local" || envLower === "development") {
          return "local";
        } else if (envLower === "production" || envLower === "prod") {
          return "production";
        } else if (envLower === "staging" || envLower === "stage") {
          return "staging";
        } else if (envLower === "hosted" || envLower === "cloud") {
          return "hosted";
        }

        // For other values, use the value as-is but sanitized
        return envLower.replace(/[^a-z0-9_-]/g, "_");
      }

      async handleLLMStart(...args: unknown[]): Promise<void> {
        logger.debug("Langfuse: LLM start intercepted");
        if (this.isVerbose) {
          logger.debug(`Langfuse: LLM start args: ${JSON.stringify(args)}`);
        }
        return super.handleLLMStart(...args);
      }

      async handleToolStart(...args: unknown[]): Promise<void> {
        logger.debug("Langfuse: Tool start intercepted");
        if (this.isVerbose) {
          logger.debug(`Langfuse: Tool start args: ${JSON.stringify(args)}`);
        }
        return super.handleToolStart(...args);
      }

      async handleRetrieverStart(...args: unknown[]): Promise<void> {
        logger.debug("Langfuse: Retriever start intercepted");
        if (this.isVerbose) {
          logger.debug(
            `Langfuse: Retriever start args: ${JSON.stringify(args)}`
          );
        }
        return super.handleRetrieverStart(...args);
      }

      async handleAgentAction(...args: unknown[]): Promise<void> {
        logger.debug("Langfuse: Agent action intercepted");
        if (this.isVerbose) {
          logger.debug(`Langfuse: Agent action args: ${JSON.stringify(args)}`);
        }
        return super.handleAgentAction(...args);
      }

      async handleAgentEnd(...args: unknown[]): Promise<void> {
        logger.debug("Langfuse: Agent end intercepted");
        if (this.isVerbose) {
          logger.debug(`Langfuse: Agent end args: ${JSON.stringify(args)}`);
        }
        return super.handleAgentEnd(...args);
      }
    }

    // Create the handler with configuration
    // Get initial metadata and tags for handler initialization
    const initialMetadata =
      metadata || (metadataProvider ? metadataProvider() : {});
    const initialTags = tagsProvider ? tagsProvider() : [];

    const config = {
      publicKey: getEnvVar("LANGFUSE_PUBLIC_KEY"),
      secretKey: getEnvVar("LANGFUSE_SECRET_KEY"),
      baseUrl:
        getEnvVar("LANGFUSE_HOST") ||
        getEnvVar("LANGFUSE_BASEURL") ||
        "https://cloud.langfuse.com",
      flushAt: Number.parseInt(getEnvVar("LANGFUSE_FLUSH_AT") || "15"),
      flushInterval: Number.parseInt(
        getEnvVar("LANGFUSE_FLUSH_INTERVAL") || "10000"
      ),
      release: getEnvVar("LANGFUSE_RELEASE"),
      requestTimeout: Number.parseInt(
        getEnvVar("LANGFUSE_REQUEST_TIMEOUT") || "10000"
      ),
      enabled: getEnvVar("LANGFUSE_ENABLED") !== "false",
      // Set trace name - can be customized via metadata.trace_name or defaults to 'mcp-use-agent'
      traceName:
        stringMetadataValue(initialMetadata, "trace_name") ||
        getEnvVar("LANGFUSE_TRACE_NAME") ||
        "mcp-use-agent",
      // Pass sessionId, userId, and tags to the handler
      sessionId: stringMetadataValue(initialMetadata, "session_id"),
      userId: stringMetadataValue(initialMetadata, "user_id"),
      tags: initialTags.length > 0 ? initialTags : undefined,
      metadata: initialMetadata || undefined,
    };

    logger.debug(
      "Langfuse handler config:",
      JSON.stringify(
        {
          traceName: config.traceName,
          sessionId: config.sessionId,
          userId: config.userId,
          tags: config.tags,
        },
        null,
        2
      )
    );

    langfuseState.handler = new LoggingCallbackHandler(
      config,
      agentId,
      metadata,
      metadataProvider,
      tagsProvider
    );
    logger.debug(
      "Langfuse observability initialized successfully with logging enabled"
    );

    // Also initialize the client for direct usage if needed
    try {
      const langfuseCore = await import("langfuse").catch(() => null);
      if (langfuseCore) {
        const { Langfuse } = langfuseCore;
        langfuseState.client = new Langfuse({
          publicKey: getEnvVar("LANGFUSE_PUBLIC_KEY"),
          secretKey: getEnvVar("LANGFUSE_SECRET_KEY"),
          baseUrl: getEnvVar("LANGFUSE_HOST") || "https://cloud.langfuse.com",
        });
        logger.debug("Langfuse client initialized");
      }
    } catch (error) {
      logger.debug(`Langfuse client initialization failed: ${error}`);
    }
  } catch (error) {
    logger.debug(`Langfuse initialization error: ${error}`);
  }
}

// Only initialize if not disabled and required keys are present
if (langfuseDisabled) {
  logger.debug(
    "Langfuse tracing disabled via MCP_USE_LANGFUSE environment variable"
  );
} else if (
  !getEnvVar("LANGFUSE_PUBLIC_KEY") ||
  !getEnvVar("LANGFUSE_SECRET_KEY")
) {
  logger.debug(
    "Langfuse API keys not found - tracing disabled. Set LANGFUSE_PUBLIC_KEY and LANGFUSE_SECRET_KEY to enable"
  );
} else {
  // Create initialization promise to ensure handlers are ready when needed
  langfuseState.initPromise = initializeLangfuse();
}

// Export getters to access the state
export const langfuseHandler = () => langfuseState.handler;
export const langfuseInitPromise = () => langfuseState.initPromise;
export { initializeLangfuse };
